src/Core/Domain/CartRule/ValueObject/CartRuleAction/CartRuleActionInterface.php

```diff
@@ -62,4 +63,6 @@ public function getPercentageDiscount(): ?PercentageDiscount;
      * @return GiftProduct|null returns null when not applicable
      */
     public function getGiftProduct(): ?GiftProduct;
+
+    public function getDiscountApplicationType(): ?DiscountApplicationType;
```

Comment:
I don't like the idea of having those methods inside this interface. And I think this interface is not well designed as it does not allow any reusability.

An action can be of many types, but that's implementation details, and methods should not be part of this interface. I think this interface should not have any method.

Eventually, those actions can have some priority as they can be added in many places. Actions can even not be linked to a cart rules, but they can be a global concept for price calculation.

```
class CartRuleActionInterface 
{
    /**
	Determines at which step this cart rule should be executed.
   */
    public function getStep(): string
}
```

- `class AddGiftProduct implement CartRuleActionInterface`



- `class ApplyDiscountPercent implement CartRuleActionInterface`
- `class EnableFreeShipping implement CartRuleActionInterface`
- `class AddGiftProduct implement CartRuleActionInterface`